File: pouring_unknown_geom/scripts/new_gp_enabled/model_selection_node_new_GP_enabled.py
#!/usr/bin/env python
import rospy
import numpy as np
import rospkg
import os
import json
import threading
import cPickle as pickle
import copy
from pouring_unknown_geom.gauss_newton_opt import opt_srv_gp_compatible
from pouring_msgs.msg import PourModelPred, PouringMsg, SelectedGPModels
from std_msgs.msg import Header
from std_srvs.srv import Trigger, TriggerResponse


class Model_selection(object):
  """docstring for Model_selection"""
  def __init__(self):
    self.model_pub = rospy.Publisher('Pour_Model_opt', PourModelPred, queue_size=1, latch=True)  #publish the model
    self.volume_solution_mode = rospy.get_param("~volume_solution_mode", "param")
    rospy.loginfo("Using the {0} method to find volume".format(self.volume_solution_mode))
    self.thread_lock = threading.Lock() # self.thread_lock.acquire(); self.thread_lock.release()
    self.setup_variables()

  def setup_variables(self, opt_num_iter=1000, opt_residual_threshold=1e-2, dom_win=[0,2.4], coef_type="power"):
    self.initial_angle = None  #this is required to shift all GP volume models down for accurate comparison
    #Pouring data used for fitting curve and updating model
    self.thread_lock.acquire()
    self.pouring_data = {'v':[], 'dv':[], 'ang':[], 'stamp': []};# self.pouring_data['dv'].append(0.0); self.pouring_data['ang'].append(0.0)
    self.collected_gp_selected_models = False
    self.thread_lock.release()
    #For obtaining gp selected models
    self.gp_selected_models_data = None
    param_coef_order = rospy.get_param("volume_profile_poly_degree",9) #This should be set...
    self.curr_coef = np.polynomial.polynomial.Polynomial(np.zeros([param_coef_order]),domain=dom_win,window=dom_win)
    self.coef_type = coef_type#"legendre" or "power"
    #Make optimizer class: #opt_num_iter: max number of iterations; #opt_residual_threshold: max change in coefficients to show convergence (break convergence loop)
    self.opt_cls = opt_srv_gp_compatible.Optimizer_class(opt_max_iter=opt_num_iter, domain_win=dom_win) #num_iter is max number of iterations to run per opt
    #subscribers
    self.sub = rospy.Subscriber('pouring_msg',PouringMsg,self.pouring_msg_callback, queue_size=1) # rospy.loginfo("subscribing")
    self.sub_gp_mod = rospy.Subscriber('gp_selected_models', SelectedGPModels, self.gp_selected_models_callback ,queue_size=1)
    self.timer = rospy.Timer(rospy.Duration(0.01), self.optimization_callback) #run this at 100hz (will be slower with opt hanging)



  def reset_callback(self, req):
    rospy.logwarn("model_selection_node Reset requested")
    self.setup_variables()
    return TriggerResponse(True,"Reset complete")

  def gp_selected_models_callback(self,data):
    self.thread_lock.acquire()
    local_bool = self.collected_gp_selected_models
    self.thread_lock.release()
    rospy.loginfo("collected gp selected models")
    if local_bool == False:
      print "collecting gp selected models in model select node"
      self.thread_lock.acquire()
      self.gp_selected_models_data = data
      self.collected_gp_selected_models = True
      self.thread_lock.release()
    else:
      pass

  def pouring_msg_callback(self,req):
    self.thread_lock.acquire()
    gp_models_obtained_local_bool = self.collected_gp_selected_models
    self.thread_lock.release()
    print "inside pouring msg callback, and collected gp: ", gp_models_obtained_local_bool
    # if gp_models_obtained_local_bool:
    if True:
      #this function takes in the msg PouringMsg which consists of a) header b) float64 dvolume (derivative volume) c) float64 angle
      rospy.loginfo("inside pouring msg callback")
      #Append this to the running list of dv and ang:
      self.thread_lock.acquire()
      self.pouring_data['v'].append(req.volume)
      self.pouring_data['dv'].append(req.dvolume)
      self.pouring_data['ang'].append(req.angle)
      self.pouring_data['stamp'].append(req.header.stamp)
      self.thread_lock.release()
      if not self.initial_angle:
        #hasn't been filled yet
        self.initial_angle = req.angle
        self.shift_gp_model_curves(init_angle=self.initial_angle)
        rospy.loginfo("Shifted gp models start volume")
    else:
      #no models have been obtained, they should be obtained even if soley parametric method is used...
      pass

  def shift_gp_model_curves(self,init_angle=None):
    #In this function, shift all of the volume curves down based on starting angle
    self.thread_lock.acquire()
    local_data_models = self.gp_selected_models_data
    self.thread_lock.release()
    for idx in range(len(local_data_models.gp_volume_models)):
      print "len of th list: ",len(local_data_models.gp_volume_models[idx].th)
      th_diff = [(th-init_angle)**2 for th in local_data_models.gp_volume_models[idx].th]
      idx_min = np.argmin(th_diff)
      V_zero = local_data_models.gp_volume_models[idx].V[idx_min]
      new_V_list = [vcurr - V_zero for vcurr in local_data_models.gp_volume_models[idx].V]
      local_data_models.gp_volume_models[idx].V = new_V_list
    self.thread_lock.acquire()
    self.gp_selected_models_data = local_data_models
    self.thread_lock.release()

  def optimization_callback(self, event):
    # print "len of pour data", len(self.pouring_data['ang']),"\n gp selected models", self.collected_gp_selected_models
    if len(self.pouring_data['ang']) > 0:
      header = Header()
      header.stamp = self.pouring_data['stamp'][-1]
      self.optimizer(header=header)

  def generate_subscribers(self):
    #get the reported dV, angle for optimization
    # self.sub = rospy.Subscriber('pouring_msg',PouringMsg,self.pouring_msg_callback, queue_size=1) # rospy.loginfo("subscribing")
    # self.sub_gp_mod = rospy.Subscriber('gp_selected_models', SelectedGPModels, self.gp_selected_models_callback ,queue_size=1)
    # self.timer = rospy.Timer(rospy.Duration(0.01), self.optimization_callback) #run this at 100hz (will be slower with opt hanging)
    self.setup_variables()
    self.reset_srv = rospy.Service('reset_model_selection_node', Trigger, self.reset_callback)
    rospy.spin()

  def optimizer(self, header=None):
    #This is running on a seperate thread from the callbacks
    #Note: keep the opt iter down so updated info can reach quickly
    if len(self.pouring_data['ang']) > 1:
      rospy.loginfo('Inside optimizer script')
      self.thread_lock.acquire()
      pouring_data = copy.copy(self.pouring_data)
      gp_volume_models = self.gp_selected_models_data.gp_volume_models
      lowest_model_rmse = self.gp_selected_models_data.lowest_model_rmse
      curr_coef = self.curr_coef #get the current coefficient estimate (this is stored and used as a seed value each round)
      self.thread_lock.release()
      #Extract vars
      v_data = copy.copy(pouring_data['v'])
      dv_data = copy.copy(pouring_data['dv'])
      theta_data = copy.copy(pouring_data['ang'])
      #Optimize (currently expects coef as legendre)
      curr_coef_upd, residuals_avg, residuals_std = self.opt_cls.update_coeff(curr_coef=curr_coef, vol_data=v_data,theta_data=theta_data)
      #Save
      self.thread_lock.acquire()
      self.curr_coef = curr_coef_upd
      self.thread_lock.release()
      if self.coef_type in "legendre":
        #Convert to legendre if desired
        domain =curr_coef_upd.domain
        curr_coef_upd = np.polynomial.legendre.Legendre.cast(curr_coef_upd, window=domain, domain=domain)
      #Unpack:
      coeff = curr_coef_upd.coef.tolist()
      domain = curr_coef_upd.domain.tolist()
      best_residual_avg =residuals_avg
      best_residual_std = residuals_std
      #Publish the current optimal model
      pour_model_msg = PourModelPred()
      pour_model_msg.header = header
      pour_model_msg.coef_opt = coeff
      pour_model_msg.domain = domain
      pour_model_msg.residual_avg = best_residual_avg
      pour_model_msg.residual_std = best_residual_std
      if self.coef_type in "legendre":
        pour_model_msg.type = 'legendre'
      else:
        pour_model_msg.type = 'power'
      #add GP model
      pour_model_msg.gp_volume_models = gp_volume_models
      pour_model_msg.lowest_model_rmse = lowest_model_rmse
      pour_model_msg.volume_solution_mode = self.volume_solution_mode #options are "param"; "semiparam"; "nonparam"
      rospy.loginfo("Publishing pour model")
      self.model_pub.publish(pour_model_msg)
      rospy.loginfo("Model published")
    else:
      rospy.loginfo('No (dv,ang) data has been recieved received')

def main():
  #This node toggles between classification and optimization and publishes the current best estimate of the container volume curve model
  rospy.init_node('model_selection_node')
  cls_obj = Model_selection()
  print "inisde model selection node"
  cls_obj.generate_subscribers()

if __name__ == '__main__':
  main()

File: pouring_unknown_geom/src/model_selection_node_new_GP_enabled.cpp
/*
 *<Monroe Kennedy III>
 *This script works to optimize coefficients and provide the model for the rest of the pipeline
 */

#include <pouring_unknown_geom/basic_ros_include.h>
#include <pouring_unknown_geom/model_selection_node_header.h>


class ModelSelectionClass
{
public:
  ModelSelectionClass(ros::NodeHandle &nh);
  ros::Publisher model_pub;
  ros::ServiceServer reset_srv;
  ros::Subscriber pouring_msg_sub;
  ros::Subscriber gp_selected_models_sub;
  ros::ServiceClient py_opt_sc;
  //pouring data record
  PouringData pouring_data;
  //params
  std::string volume_solution_mode; //taken from param server: volume_solution_mode
  int param_coef_order; //get from param: volume_profile_poly_degree
  //vars
  std::vector<double> curr_coef; //current list of coefficients
  std::vector<double> coef_domain;
  std::string coef_type;
  double initial_angle; //initial angle of onset of pouring (used to shift volume profiles for training data)
  pouring_msgs::SelectedGPModels gp_selected_models;


  void GPSelectedModelsCallback(const pouring_msgs::SelectedGPModels &data);
  void PouringMsgCallback(const pouring_msgs::PouringMsg &data);
  bool ResetCallback(std_srvs::Trigger::Request &req, std_srvs::Trigger::Response &resp);
  void ResetFunction();
  void MainLoop(); //runs main loop at specified hz in main func
private:
  ros::NodeHandle pnh_;
  //Flags:
  bool collected_gp_selected_models;
  bool init_angle_set_bool;
  bool volume_data_shifted_bool;
  int pour_model_msg_index_counter;

};

ModelSelectionClass::ModelSelectionClass(ros::NodeHandle &nh) : pnh_(nh)
{
  reset_srv = pnh_.advertiseService("reset_model_selection_node",&ModelSelectionClass::ResetCallback,this); //reset trigger
  pouring_msg_sub = pnh_.subscribe("pouring_msg",1, &ModelSelectionClass::PouringMsgCallback,this); //pouring msg for volume/angle data
  gp_selected_models_sub = pnh_.subscribe("gp_selected_models",1,&ModelSelectionClass::GPSelectedModelsCallback,this); // gp selected models (is simply passed through)
  model_pub = pnh_.advertise<pouring_msgs::PourModelPred>("Pour_Model_opt",1, true); //publish the model
  //Setup the optimization service client
  ros::service::waitForService("py_profile_opt_srv");
  py_opt_sc = pnh_.serviceClient<pouring_msgs::PyOptimizationServ>("py_profile_opt_srv");
  //Call the setup

  collected_gp_selected_models = false; //only needs to be set once
  coef_domain.push_back(0.0);
  coef_domain.push_back(2.4);
 
  pour_model_msg_index_counter = 0;
  ModelSelectionClass::ResetFunction();
}


void ModelSelectionClass::ResetFunction()
{
  //Reset key variables
  //Call params
  pnh_.param<std::string>("volume_solution_mode",volume_solution_mode,"param");
  pnh_.param<int>("volume_profile_poly_degree",param_coef_order,9);
  //populate the curr coef
  curr_coef = std::vector<double>();
  for(int idx=0; idx<param_coef_order; ++idx)
  {
    curr_coef.push_back(0.0);
  }
  //set default domain
  pnh_.param<std::string>("volume_profile_polynomial_type",coef_type, "power");
  //Reset variables
  pouring_data =  PouringData();
  init_angle_set_bool = false;
  //collected_gp_selected_models = false;  //Unneccessary, as if a new container is pub then this will update
  volume_data_shifted_bool = false; //shifts vol data wrt start angle
  // gp_selected_models = pouring_msgs::SelectedGPModels(); //Don't reset this in case latch topic doesn't echo again, then you have something
}

void ModelSelectionClass::GPSelectedModelsCallback(const pouring_msgs::SelectedGPModels &data)
{
  //Populate gp selected models variable with latest
  if(collected_gp_selected_models==false){
    gp_selected_models = data;
    collected_gp_selected_models = true;
  }
}


void ModelSelectionClass::PouringMsgCallback(const pouring_msgs::PouringMsg &data)
{
  //Populate the variable pouring_data
  pouring_data.v.push_back(data.volume); 
  pouring_data.dv.push_back(data.dvolume); 
  pouring_data.ang.push_back(data.angle);
  pouring_data.stamp.push_back(data.header.stamp);
  //Store initial angle
  if(init_angle_set_bool == false){
    //Store and shift the angle of the models once obtained
    initial_angle = data.angle;
    init_angle_set_bool = true;
  }
  if(init_angle_set_bool==true && volume_data_shifted_bool==false && collected_gp_selected_models==true)
  {
    //shift the volume data
    ShiftVolumeDataFunct(gp_selected_models, initial_angle);//impl in header
    volume_data_shifted_bool = true;
  }
}


bool ModelSelectionClass::ResetCallback(std_srvs::Trigger::Request &req, std_srvs::Trigger::Response &resp)
{
  //Reset necessary params here
  ModelSelectionClass::ResetFunction();
  resp.success = true;
  resp.message = "model_selection_node reset successfully";
  return true;
}


void ModelSelectionClass::MainLoop()
{
  //1. Get PourMsg (only proceed if more than 2 readings are present) [V,th] list
  //ROS_INFO("Inside top main loop");
  //std::cout << " collected gp models" << collected_gp_selected_models << "\n pouring data size "<< pouring_data.v.size() << std::endl;
  if(collected_gp_selected_models==true && pouring_data.v.size() > 1){
    //gp models have been collected and more than 2 volume data points have been stored [and therefore gp selected models have also been shifted]
    //ROS_INFO("conditions met for entering if statement of main loop");
    //2. Get the latest values and coef
    pouring_data;
    gp_selected_models;
    curr_coef;

    //3. Construct the optimization message
    pouring_msgs::PyOptimizationServ py_opt_req;
    for(int idx=0; idx<pouring_data.v.size(); ++idx)
    {
      py_opt_req.request.vol_data.push_back(pouring_data.v[idx]);
      py_opt_req.request.ang_data.push_back(pouring_data.ang[idx]);
    }
    py_opt_req.request.coef_init = curr_coef;

    if (py_opt_sc.call(py_opt_req))
    {
      //ROS_INFO("py_opt_sc callback successful");
      //Now send the update to the publisher and publish
      pouring_msgs::PourModelPred pour_model_msg;
      pour_model_msg.header.seq = pour_model_msg_index_counter; pour_model_msg_index_counter += 1;
      pour_model_msg.header.stamp = ros::Time::now();
      //Coef params
      pour_model_msg.coef_opt = py_opt_req.response.coef_opt;
      pour_model_msg.domain = coef_domain;
      pour_model_msg.residual_avg = py_opt_req.response.resid_avg;
      pour_model_msg.residual_std = py_opt_req.response.resid_std;
      pour_model_msg.type= coef_type;
      //Add the GP models
      pour_model_msg.gp_volume_models = gp_selected_models.gp_volume_models;
      pour_model_msg.volume_solution_mode = volume_solution_mode;
      //Publish
      model_pub.publish(pour_model_msg);
    }
    else
    {
      ROS_WARN("py_opt_sc callback NOT successful");
    }
  }//end olf if statement for collected models and pouring size check
}


int main(int argc, char **argv)
{
  ros::init(argc, argv, "model_selection_node");
  ros::NodeHandle nh("~");
  ModelSelectionClass cls_obj(nh);
  ros::Rate loop_rate(100); //HZ
  while(ros::ok()){
    ros::spinOnce();
    cls_obj.MainLoop();
    loop_rate.sleep();
  }

  return 0;
}
